stdlib/public/core/Array.swift

```diff
@@ -1865,3 +1865,187 @@ internal struct _ArrayAnyHashableBox<Element: Hashable>
     return true
   }
 }
+
+// SWIFT_ENABLE_TENSORFLOW
+// `Array<Element>` is a countably infinite direct sum 
+// (https://en.wikipedia.org/wiki/Direct_sum) of `Element`, when `Element` 
+// conforms to `AdditiveArithmetic` and thus has a zero.
+//
+// Therefore, the most useful (co)tangent space for `Array<Element>` is the 
+// countably infinite direct sum of 
+// `Element.TangentVector`/`Element.CoTangentVector` (since 
+// `Element.TangentVector` and `Element.CoTangentVector` conform to 
+// `AdditiveArithmetic`, they have a zero). If we think of all the elements past 
+// the end of an array as zeros, values of these direct sums can be represented 
+// by values of `Array<Element.TangentVector>`/`Array<Element.CotangentVector>`. 
+// 
+// The operations are:
+// - `zero` is `[]` (which is equivalent to `[.zero]`, `[.zero, .zero]`, etc).
+// - `+` is (pseudocode): 
+//   ```
+//   a + b = range(max(a.count, b.count)).map {
+//     i in (a[i] or .zero) + (b[i] or .zero)
+//   }
+//   ```
+// 
+// This leads to expected behavior when differentiating functions on arrays. 
+// For example:
+// ```
+// func sumFirstThree(_ array: Array<Float>) -> Float {
+//   return array[0] + array[1] + array[2]
+// }
+// ```
+// We expect the gradient of that to be `[1, 1, 1]`. That is what we get, if we 
+// define the pullback of array subscripting as follows:
+// ```
+// func vjpSubscript(index) {
+//   let result = self[index]
+//   let pullback = { v in
+//     var pb = Array(repeating: zero, count: index + 1)
+//     pb[index] = v
+//     return pb
+//   }
+//   return (result, pullback)
+// }
+// ```
+//
+extension Array : Differentiable where Element : Differentiable {
+  public typealias AllDifferentiableVariables = [Element]
+
+  public struct TangentVector : AdditiveArithmetic & Differentiable {
+    public typealias AllDifferentiableVariables = Array.TangentVector
+    public typealias TangentVector = Array.TangentVector
+    public typealias CotangentVector = Array.CotangentVector
+
+    public var elements: [Element.TangentVector]
+
+    public static var zero: Array.TangentVector {
+      return TangentVector(elements: [])
+    }
+    
+    public init(elements: [Element.TangentVector]) {
+      self.elements = elements
+    }
+
+    @inlinable
+    public static func + (
+      lhs: Array.TangentVector,
+      rhs: Array.TangentVector
+    ) -> Array.TangentVector {
+      let zippedElements = zipLongest(lhs.elements, rhs.elements)
+      return Array.TangentVector(
+        elements: zippedElements.map { ($0 ?? .zero) + ($1 ?? .zero) })
+    }
+
+    @inlinable
+    public static func - (
+      lhs: Array.TangentVector,
+      rhs: Array.TangentVector
+    ) -> Array.TangentVector {
+      let zippedElements = zipLongest(lhs.elements, rhs.elements)
+      return Array.TangentVector(
+        elements: zippedElements.map { ($0 ?? .zero) - ($1 ?? .zero) })
+    }
+
+    @inlinable
+    public func moved(
```

Comment:
These implementations of `moved` and `tangentVector` are slightly wrong because they throw away elements from the longer array. For example,
```
[1, 2].moved(along: [1, 2, 3]) = [2, 4] // the `3` in the direction got thrown away
```

There's no way to define a reasonable `moved` that never throws away elements of `direction`! I suggest:
* adding a precondition that `self` is at least as long as `direction`, so that there is no silent throwing away
* making it keep all the elements of `self` when `self` is longer than `direction`.

Similar comment applies to `tangentVector`.

Reply:
Good point! Given our treatment of arrays as countably infinite. direct sums, wouldn't it be ok to use `zipLongest` and allow `move` and `tangentVector` to return arrays that are longer than `self`, if `direction` is longer?

Reply:
The problem is that `Self` doesn't necessarily have a `.zero` to use to fill in the missing elements.

In the case where `Self: AdditiveArithmetic`, what you say is doable. Maybe it would be possible to add an implementation of this method that does what you say when `Self: AdditiveArithemetic`. And fall back to the implementation with the precondition otherwise. Not sure if there is a way to make the type system do this.

Reply:
Oh right. I was actually curious about that. Why do we not want `Differentiable` to only apply to `AdditiveArithmetic` types? What is an example of a differentiable type that does not conform to `AdditiveArithmetic`?

Reply:
> Why do we not want `Differentiable` to only apply to `AdditiveArithmetic` types? What is an example of a differentiable type that does not conform to `AdditiveArithmetic`?

Good question. From the [Differentiable Types](https://github.com/tensorflow/swift/blob/test/docs/DifferentiableTypes.md#protocol-details) document:

> Mathematically, `Differentiable` represents a differentiable manifold: this is a technical term for smooth-surfaced objects like spheres and generalizes types that are compatible with differentiation, like `Float`, `Double`, `Tensor`, and `SIMD4<Float>`.

I don't believe addition is defined for differentiable manifolds in general: take a sphere, for instance. On the other hand, tangent vectors (i.e. elements of the tangent space of a differentiable manifold) do define addition.

Reply:
See our new [Differentiable Types](https://github.com/tensorflow/swift/blob/master/docs/DifferentiableTypes.md) document for answers. A short answer is that 1. general differentiable manifolds are not additive groups, and 2. differentiable structs with non-differentiable (`@noDerivative`) members are not additive groups.

Reply:
I have two examples.

A real situation that happens in actual layers and models that we have made:
```
struct Model: Differentiable {
  @noDerivative var flag: Bool  // some flag that determines some behavior
  var parameters: Tensor<Float>
}
```
We want to be able to differentiate wrt `Model` to optimize it. The gradient says how `parameters` should change. There is no reasonable way to make it `AdditiveArithmetic` because there is no good "zero" or "+" that does reasonable things with `flag`.

A more abstract example: Suppose you have a function `f: Sphere -> Float`. It makes sense to differentiate it. The gradient is some vector in the cotangent space of `Sphere`. There is no "zero" or "+" on `Sphere`.

Reply:
Thanks for all the great examples/explanations! :)